F0 usart

Jeff-Ciesielski opened the pull request:
This patch set adds an stm32f0 interrupt driven usart driver, as well as a serial command shell and stm32f0 example

Jeff-Ciesielski committed:
Add STM32F0 usart driver and command shell

drivers are now encapsulated as a library, this pattern should be
followed for other MCUs

Jeff-Ciesielski committed:
Add f0 serial communications example

Jeff-Ciesielski committed:
Add f0 serial communications example

vadimplobanov commented on Makefile:
```diff
@@ -82,19 +82,23 @@ ifneq ($(V),1)
-APP_D_FILES            = $(APP_C_FILES:.c=.d)
-APP_O_FILES            = $(APP_C_FILES:.c=.o)            $(APP_S_FILES:.s=.o)
+STM32F0_DRIVERS_D_FILES = $(STM32F0_DRIVERS_C_FILES:.c=.d)
+STM32F0_DRIVERS_O_FILES = $(STM32F0_DRIVERS_C_FILES:.c=.o)
```
Maybe do the .s -> .o rule here as well? Even if you have no assembly now, it keeps people from being surprised later should they happen to add some. ("Why aren't these files getting built? What the heck?") Make sure to init STM32F0_DRIVERS_S_FILES to empty, though.

vadimplobanov commented on Makefile:
```diff
@@ -82,19 +82,23 @@ ifneq ($(V),1)
+APP_D_FILES             = $(APP_C_FILES:.c=.d)
+APP_O_FILES             = $(APP_C_FILES:.c=.o)            $(APP_S_FILES:.s=.o)
+SHELL_D_FILES           = $(SHELL_C_FILES:.c=.d)
+SHELL_O_FILES           = $(SHELL_C_FILES:.c=.o)
```
Ditto as above.

vadimplobanov commented on Makefile:
```diff
@@ -140,6 +152,10 @@ libstm32_usb.a: $(STM32_USB_O_FILES)
 clean:
 	$(Q)rm -f \
+	$(SHELL_O_FILES)            \
+	$(SHELL_D_FILES)            \
```
OCD trigger. :) Maybe keep these ordered exactly as above, including _D_FILES before _O_FILES?

vadimplobanov commented on Makefile:
```diff
@@ -168,6 +184,7 @@ ddd: $(TARGET).elf
 gdb: $(TARGET).elf
 	$(GDB) -ex "target ext localhost:3333" -ex "mon reset halt" -ex "mon arm semihosting enable" $(TARGET).elf
 
+-include $(SHELL_D_FILES)
```
Also ordering? And do we need to include the STM32F0_DRIVERS_D_FILES as well? If not, why not?

vadimplobanov commented on app/f0_serialcom/FreeRTOSConfig.h:
```diff
@@ -0,0 +1,88 @@
+/*
```
This file's copied from FreeRTOS distributables, right?

Jeff-Ciesielski replied:
Yes, copied and modified accordingly

vadimplobanov replied:
Ok, cool, that's why I didn't comment on the style. Figured this was the case from the license comment block.

vadimplobanov commented on app/f0_serialcom/blinky.h:
```diff
@@ -0,0 +1,5 @@
+#define _BLINKY_H_
+
+void blinky_task(void *params);
+#endif	/* BLINKY_H */
```
Have an empty line between the function declarations and the endif, maybe? /bikeshed-bikeshed-bikeshed :)

vadimplobanov commented on app/f0_serialcom/main.c:
```diff
@@ -0,0 +1,97 @@
+{
+}
+
+void vApplicationStackOverflowHook(xTaskHandle pxTask, signed char *pcTaskName)
```
Are you sure about the type of pcTaskName? I know the type is mandated by FreeRTOS, but could you double check that it's really a signed char pointer? I'd have expected a bare-word char...

Jeff-Ciesielski replied:
Yep, it's a signed char

vadimplobanov replied:
That's just... weird. Makes me think that the person who wrote this part of FreeRTOS doesn't know this part of the C standard.

vadimplobanov commented on app/f0_serialcom/main.c:
```diff
@@ -0,0 +1,97 @@
+void HardFault_Handler(void)
+{
+	while(1)
+		;
```
Perfect opportunity to enable disco-mode with the LEDs here. :D (kidding, kidding...)

vadimplobanov commented on app/f0_serialcom/main.c:
```diff
@@ -0,0 +1,97 @@
+	char *tbuf = alloca(task_count * 40);
+	printf("Name         State     Prio    Stack  Num\n");
+	printf("-----------------------------------------\n");
+	vTaskList((signed char*)tbuf);
```
Same signed-char curiosity here. Reason why I mention it is that 'char' != 'signed char', despite the fact that the equality does hold for all non-char integral types. (C is weird that way...)

vadimplobanov commented on app/f0_serialcom/main.c:
```diff
@@ -0,0 +1,97 @@
+static void tasks_handler(int argc, char **argv)
+{
+	int task_count = uxTaskGetNumberOfTasks();
+	char *tbuf = alloca(task_count * 40);
```
Is it possible to do anything to protect ourselves from smashing our own stack, in case the task names are longer than 40? Perhaps a MAX_TASK_NAME define, or something similar, that would come in handy?

vadimplobanov commented on app/f0_serialcom/main.c:
```diff
@@ -0,0 +1,97 @@
+	(void)params;
+
+	/* Unbuffered standard IO */
+	setvbuf(stdin, NULL, _IONBF, 0);
```
OCD idea: put extra space after "stdin," to make it all line up nicely?

vadimplobanov commented on app/f0_serialcom/main.c:
```diff
@@ -0,0 +1,97 @@
+	
+	shell_init();
+
+	shell_register_command(&tasks_cmd);
```
Could use a little comment that shell_register_command will copy its parameter and therefore this is safe, so that folks don't get worried about long-term references to a stack variable.

vadimplobanov commented on app/f0_serialcom/main.c:
```diff
@@ -0,0 +1,97 @@
+
+	shell_register_command(&tasks_cmd);
+	xTaskCreate(blinky_task,
+		    (signed portCHAR *) "Blink",
```
Hmm, here we use a signed-portCHAR as opposed to a signed-char for task names. Could use some unification on the type of the task names.

Jeff-Ciesielski replied:
Yeah, I see what you're saying.  This is copied pretty much verbatim from their examples, but I'll change to a signed char.

vadimplobanov commented on app/f0_serialcom/main.c:
```diff
@@ -0,0 +1,97 @@
+		    (signed portCHAR *) "Blink",
+		    configMINIMAL_STACK_SIZE,
+		    NULL,
+		    (tskIDLE_PRIORITY + 2),
```
Unneeded parens?

vadimplobanov commented on app/f0_serialcom/main.c:
```diff
@@ -0,0 +1,97 @@
+
+void HardFault_Handler(void)
+{
+	while(1)
```
Space between while and paren? Maybe even "for (;;)" instead?

vadimplobanov commented on app/f0_serialcom/main.c:
```diff
@@ -0,0 +1,97 @@
+	vTaskDelete(NULL);
+
+	/* Shouldn't ever get here */
+	while(1);
```
Ditto.

vadimplobanov commented on app/f0_serialcom/main.c:
```diff
@@ -0,0 +1,97 @@
+int main(void)
+{
+	xTaskCreate(setup_task,
+		    (signed portCHAR *) "Setup",
```
Ditto.

vadimplobanov commented on app/f0_serialcom/main.c:
```diff
@@ -0,0 +1,97 @@
+		    (signed portCHAR *) "Setup",
+		    configMINIMAL_STACK_SIZE,
+		    NULL,
+		    (tskIDLE_PRIORITY + 2),
```
Ditto.

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+#include <shell.h>
+
+char *shell_prompt_text = "> ";
+char *shell_banner = "\n\nCommand Shell";
```
Both of these want to be static const char *.

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+char *shell_prompt_text = "> ";
+char *shell_banner = "\n\nCommand Shell";
+
+struct shell_state {
```
This wants to be static. If it looks weird that way, then perhaps split the struct declaration from the variable declaration.

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+struct shell_state {
+	/* +1 for the help command */
+	struct shell_cmd commands[SHELL_MAX_COMMANDS + 1];
+	int command_count;
```
unsigned int

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+	struct shell_cmd commands[SHELL_MAX_COMMANDS + 1];
+	int command_count;
+	char input_buf[32];
+	int input_pos;
```
unsigned int

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+	xSemaphoreHandle shell_lock;
+} shell_state;
+
+static bool shell_is_whitespace(char c)
```
I'm guessing that neither FreeRTOS nor gcc give you a working ctype.h, so you have to roll your own?

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+
+static bool shell_is_whitespace(char c)
+{
+	switch(c) {
```
space between switch and paren

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+	case '\r':
+	case '\n':
+		return true;
+	default:
```
Maybe even get rid of the default label, since it'll fall through?

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+{
+	unsigned int cmd_len;
+	unsigned int tgt_cmd_len = strlen(argv[0]);
+	int i;
```
unsigned int

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+
+static void shell_exec_cmd(int argc, char **argv)
+{
+	unsigned int cmd_len;
```
Push variable declaration into the body of the for loop?

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+		cmd_len = strlen(shell_state.commands[i].command);
+
+		if (cmd_len == tgt_cmd_len && shell_state.commands[i].action &&
+		    strncmp(argv[0], shell_state.commands[i].command, tgt_cmd_len) == 0) {
```
Since the length is already known to match, why not do a simple strcmp()?

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+	unsigned int tgt_cmd_len = strlen(argv[0]);
+	int i;
+	for (i = 0; i < shell_state.command_count; i++) {
+		cmd_len = strlen(shell_state.commands[i].command);
```
The body of the loop might look cleaner if you assign shell_state.commands[i] to a const struct shell_cmd * local variable.

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+
+static void shell_parse_command(void)
+{
+	int i;
```
unsigned int

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+static void shell_parse_command(void)
+{
+	int i;
+	int arg_count = 0;
```
unsigned int

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+{
+	int i;
+	int arg_count = 0;
+	char *arg_list[8] = {NULL};
```
Can get away without initializing this array, since it's protected by the count variable above.

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+	bool in_arg = false;
+
+	/* TODO: Fix the bug where I pass everything as args...*/
+	for(i = 0; i < shell_state.input_pos; i++) {
```
Space between for and paren.

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+	char input_buf[32];
+	int input_pos;
+	bool initialized;
+	xSemaphoreHandle shell_lock;
```
Perhaps call it registration_lock, since it only seems to protect from concurrent shell command registrations?

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+	int command_count;
+	char input_buf[32];
+	int input_pos;
+	bool initialized;
```
Seems that this gets set at init and then never used.

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+
+	/* TODO: Fix the bug where I pass everything as args...*/
+	for(i = 0; i < shell_state.input_pos; i++) {
+		if (shell_is_whitespace(shell_state.input_buf[i]) && arg_count == 0)
```
I think this entire conditional is subsumed by the code below, and is superfluous.

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+
+		if (shell_is_whitespace(shell_state.input_buf[i])) {
+			if (in_arg) {
+				shell_state.input_buf[i] = 0x00;
```
We're dealing with text, so perhaps '\0' might be more appropriate than 0x00?

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+				in_arg = false;
+			}
+		} else if (!in_arg) {
+			arg_list[arg_count] = &shell_state.input_buf[i];
```
You need to protect yourself from running off the end of the arg_list.

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+			in_arg = true;
+		}
+	}
+	shell_state.input_buf[i] = 0x00;
```
If the entire input_buf was used for the command, then you just scribbled past the end of it. You need to give yourself room for one additional character in order to be able to nil-terminate the final arg in case of a completely full buffer.

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+		shell_exec_cmd(arg_count, arg_list);
+}
+
+static int shell_process_input(void)
```
Nothing seems to care about this function's return value.

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+{
+	char c;
+
+	c = (char)getchar();
```
Can you ever get EOF (a negative value) here?

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+
+	c = (char)getchar();
+
+	switch(c) {
```
Ditto.

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+	case '\n':
+		putchar('\n');
+		/* handle command here */
+		shell_parse_command();
```
If you ever get "hello world\r\n" as input, then you're going to process it as two commands ("hello world" and ""). Easy way to handle this case is to only call shell_parse_command() if input_pos != 0.

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+		/* handle command here */
+		shell_parse_command();
+
+		memset(shell_state.input_buf, 0x00, ARRAY_SIZE(shell_state.input_buf));
```
Don't think you need this, since input_pos is set to zero just below and therefore marks the entire buffer as "don't look" anyway.

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+			shell_state.input_pos--;
+			putchar('\b');
+			putchar(' ');
+			putchar('\b');
```
I'm curious: why a bell, a space, and another bell? :)

vadimplobanov replied:
I'm curious: why a backspace, a space, and another backspace? :)

Jeff-Ciesielski replied:
VT100.  Moves cursor back, blanks the character (which advances the cursor), then moves the cursor back again.

vadimplobanov replied:
Okay, gotcha!

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+		break;
+	default:
+		/* we've overrun our buffer */
+		if ((unsigned)shell_state.input_pos >= ARRAY_SIZE(shell_state.input_buf)) {
```
If input_pos is unsigned, the cast becomes unnecessary.

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+	default:
+		/* we've overrun our buffer */
+		if ((unsigned)shell_state.input_pos >= ARRAY_SIZE(shell_state.input_buf)) {
+			printf("Invalid command!\n");
```
It might be better for usability if you instead remember the fact that the command overran its buffer and do this printout when you get the line ending (\r or \n). That way the user gets an error when they hit enter, rather than in the middle of typing. In the latter case, the input and output is going to get jumbled together.

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+		if ((unsigned)shell_state.input_pos >= ARRAY_SIZE(shell_state.input_buf)) {
+			printf("Invalid command!\n");
+			shell_state.input_pos = 0;
+			memset(shell_state.input_buf, 0x00, ARRAY_SIZE(shell_state.input_buf));
```
Ditto.

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+
+	printf("Shell commands: (case sensitive!)\n");
+
+	for(i = 0; i < ARRAY_SIZE(shell_state.commands); i++){
```
Space before '(' and another before '{'?

vadimplobanov commented on app/f0_serialcom/leds.c:
```diff
@@ -0,0 +1,63 @@
+	gpio_conf.GPIO_Mode = GPIO_Mode_OUT;
+	gpio_conf.GPIO_OType = GPIO_OType_PP;
+
+	for (i = 0; i < 4; ++i){
```
) { perhaps?

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+			printf("%s\n", shell_state.commands[i].command);
+		} else {
+			/* This is the end of the command array*/
+			return;
```
Someone who's silly enough to register a shell command with a name that starts with '\0' can make the help text of all subsequent commands inaccessible... but you will let the user run those commands regardless. It's better to unify the help implementation with the command invocation implementation -- one easy way to do so is to just drop this else condition entirely.

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+	}
+}
+
+int shell_register_command(struct shell_cmd *cmd)
```
const struct shell_cmd?

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+	for (i = 0; i < shell_state.command_count; i++) {
+		cmd_len = strlen(shell_state.commands[i].command);
+
+		if (cmd_len == tgt_cmd_len && shell_state.commands[i].action &&
```
I think you already protect against a NULL action at command registration time.

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+void shell_process(void* params)
+{
+	printf("%s\n%s", shell_banner, shell_prompt_text);
+	while(1) {
```
Ditto.

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+	shell_register_command(&help_cmd);
+
+	xTaskCreate(shell_process,
+		    (signed portCHAR *) "Shell",
```
Ditto.

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+		    (signed portCHAR *) "Shell",
+		    configMINIMAL_STACK_SIZE + 100,
+		    NULL,
+		    (tskIDLE_PRIORITY + 2),
```
Ditto.

vadimplobanov commented on libs/shell/shell.c:
```diff
@@ -0,0 +1,209 @@
+	static struct shell_cmd help_cmd = {
+		.command = "help",
+		.action = shell_help_handler,
+		.help = shell_help_help, /* This name is stupid */
```
There's no helping some users... ;)

vadimplobanov commented on libs/shell/shell.h:
```diff
@@ -0,0 +1,15 @@
+#define SHELL_MAX_COMMANDS 10
+
+void shell_init(void);
+void shell_process(void *params);
```
Does anyone else need to know about this function? Could remove it from the header and make it static, I think.

Jeff-Ciesielski replied:
Missed that.  Shell process used to be initialized directly, but I rolled it into init.  I'll remove that.

vadimplobanov commented on libs/shell/shell.h:
```diff
@@ -0,0 +1,15 @@
+void shell_init(void);
+void shell_process(void *params);
+int shell_register_command(struct shell_cmd *command);
+#endif
```
Blank line above?

vadimplobanov commented on libs/shell/shell.h:
```diff
@@ -0,0 +1,15 @@
+
+struct shell_cmd {
+	char *command;
+	void(*action)(int argc, char **argv);
```
Maybe a space after the return type on these two?

vadimplobanov commented on libs/stm32f0_drivers/usart.c:
```diff
@@ -0,0 +1,295 @@
+		void (*tx_clock_cmd)(uint32_t, FunctionalState);
+		uint32_t tx_pin_clock;
+		uint8_t         tx_pinsource;
+	} gpios;
```
The rx and tx definitions above match. Perhaps a good reason to pull them into a struct definition?

vadimplobanov commented on libs/stm32f0_drivers/usart.c:
```diff
@@ -0,0 +1,295 @@
+	} gpios;
+
+	bool rx_overflow;
+	uint32_t irq_no;
```
The order of these two member variables differs from the initialization down below.

vadimplobanov commented on libs/stm32f0_drivers/usart.c:
```diff
@@ -0,0 +1,295 @@
+
+static struct usart *usart_lookup(uint8_t usart_no)
+{
+
```
USART1 vs USART_1 and USART2 vs USART_2 can be a little dicey. I'd recommend to make the two definitions (periph vs usart_no) be a bit more different, to avoid "oopsie typo" moments.

vadimplobanov commented on libs/stm32f0_drivers/usart.c:
```diff
@@ -0,0 +1,295 @@
+{
+	portBASE_TYPE q_ret;
+	int ret = 0;
+	static char linefeed = '\r';
```
Better to make this non-global (non-static)?

vadimplobanov commented on libs/stm32f0_drivers/usart.c:
```diff
@@ -0,0 +1,295 @@
+	if (!u)
+		return -ENODEV;
+
+	if(!u->initialized)
```
Space after if?

vadimplobanov commented on libs/stm32f0_drivers/usart.c:
```diff
@@ -0,0 +1,295 @@
+
+	while (len--) {
+		if (*s == '\n' && u->autocrlf) {
+			q_ret = xQueueSend(u->buffers.tx, &linefeed, portMAX_DELAY);
```
How about q_ret = xQueueSend(&linefeed) & xQueueSend(s)?

vadimplobanov commented on libs/stm32f0_drivers/usart.c:
```diff
@@ -0,0 +1,295 @@
+		if (q_ret != pdTRUE) {
+			return -EIO;
+		} else {
+			USART_ITConfig(u->periph, USART_IT_TXE, ENABLE);
```
I think it'll be more readable to yank this out of the else, and delete the else. The if branch will return anyway, so no reason to indent after it.

vadimplobanov commented on libs/stm32f0_drivers/usart.c:
```diff
@@ -0,0 +1,295 @@
+	if (!u)
+		return -ENODEV;
+
+	if(!u->initialized)
```
Ditto.

vadimplobanov commented on libs/stm32f0_drivers/usart.c:
```diff
@@ -0,0 +1,295 @@
+		return -ENOTCONN;
+
+	while (len--) {
+		xQueueReceive(u->buffers.rx, d++, portMAX_DELAY);
```
xQueueSend() can fail, hence why you check it with q_ret above. Can xQueueReceive() fail as well?

vadimplobanov commented on libs/stm32f0_drivers/usart.c:
```diff
@@ -0,0 +1,295 @@
+	return usart_read(usart_no, c, 1);
+}
+
+void usart_common_irq(struct usart *u)
```
This function wants to be static.

vadimplobanov commented on libs/stm32f0_drivers/usart.c:
```diff
@@ -0,0 +1,295 @@
+		uint8_t         tx_pinsource;
+	} gpios;
+
+	bool rx_overflow;
```
This variable can be set, but is never cleared or checked as far as I can see.

Jeff-Ciesielski replied:
It's a watermark for debugging. Not much to be done (though I suppose I can add a method to check this if the situation were to arise when that info were useful)

vadimplobanov commented on libs/stm32f0_drivers/usart.c:
```diff
@@ -0,0 +1,295 @@
+	}
+	portEND_SWITCHING_ISR(tx_task_woken | rx_task_woken);
+}
+
```
Where do the two functions below get referenced?

vadimplobanov commented on libs/stm32f0_drivers/usart.h:
```diff
@@ -0,0 +1,19 @@
+int usart_read(int usart_no, char *d, int len);
+int usart_putchar(int usart_no, char c);
+int usart_getchar(int usart_no, char *c);
+#endif /* _DEBUG_USART_H_ */
```
Ditto.

vadimplobanov commented on util/util.h:
```diff
@@ -0,0 +1,6 @@
+#ifndef _UTIL_H_
+#define _UTIL_H_
+
+#define ARRAY_SIZE(x) (sizeof(x) / sizeof(x[0]))
```
(sizeof(x) / sizeof((x)[0]), if you want to be extra-safe

vadimplobanov commented:
Looks awesome! Most of the comments are for tiny and very bikesheddy stuff. :)

Jeff-Ciesielski commented:
Thanks for taking the time to review.  A lot of the inconsistency is due to both the usart and shell being pulled from other past projects, but I'm glad you're calling out what you see, I'd rather have it all fit together nicely :)

vadimplobanov commented:
No problem. I erred on the side of providing more comments rather than fewer, figuring you can ignore the really nit-picky stuff if the code looks better without.

Jeff-Ciesielski committed:
Fix up usart driver, merge upstream changes
Merge branch 'f0_usart' of github.com:autosportlabs/STM32Fx_FreeRTOS_Base into f0_usart

Conflicts:
	app/f0_serialcom/blinky.h
	app/f0_serialcom/config.mk
	app/f0_serialcom/main.c
